keps/sig-node/3953-node-resource-hot-plug/README.md

```diff
@@ -0,0 +1,674 @@
+- Creating new nodes takes more time compared to increasing the capacity of existing nodes.
+
+### Goals
+
+* Dynamically scale up the node by hot plugging  resources and without restarting the kubelet.
+* Ability to reinitialize resource managers (CPU manager, memory manager) to adopt changes in node's resource.
+
+### Non-Goals
+
+* Dynamically adjust system reserved and kube reserved values.
+* Hot unplug of node resources.
+* Update the autoscaler to utilize resource hot plugging.
+
+
+## Proposal
+
+This KEP aims to support the node resource hot plugging by adding a polling mechanism in kubelet to fetch the machine-information from cAdvisor's cache which is already updated periodically, This information will be fetched periodically by kubelet, after which the node status updater is responsible for updating this information at node level in the cluster.
+
+Additionally, this KEP aims to improve the initialization and reinitialization of resource managers, such as the memory manager and CPU manager, so that they can adapt to change in node's configurations.
+
+### User Stories
+
+#### Story 1
+
+As a cluster admin, I must be able to increase the cluster resource capacity without adding a new node to the cluster.
+
+#### Story 2
+
+As a cluster admin, I must be able to increase the cluster resource capacity without need to restarting the kubelet.
+
+### Notes/Constraints/Caveats (Optional)
+
+### Risks and Mitigations
+
+1. Node resource hot plugging is an opt-in feature, merging the
+   feature related changes won't impact existing workloads. Moreover, the feature
+   will be rolled out gradually, beginning with an alpha release for testing and
+   gathering feedback. This will be followed by beta and GA releases as the
+   feature matures and potential problems and improvements are addressed.
+2. Though the node resource is updated dynamically, the dynamic data is fetched from cAdvisor and its well integrated with kubelet.
+3. Resource manager are updated to adapt to the dynamic node reconfigurations, Enough tests should be added to make sure its not affecting the existing functionalities.
+
+## Design Details
+
+
+Below diagram is shows the interaction between kubelet and cAdvisor.
+
+
+```mermaid
+sequenceDiagram
+    participant node
+    participant kubelet
+    participant cAdvisor-cache
+    participant machine-info
+    kubelet->>cAdvisor-cache: Poll
+    cAdvisor-cache->>machine-info: fetch
+    machine-info->>cAdvisor-cache: update
+    cAdvisor-cache->>kubelet: update
+    kubelet->>node: node status update
+    kubelet->>node: re-initialize resource managers
+```
+
+The interaction sequence is as follows
+1. Kubelet will be polling in interval to fetch the machine resource information from cAdvisor's cache, Which is currently updated every 5 minutes.
+2. Kubelet's cache will be updated with the latest machine resource information.
```

Comment:
what about cases when cadvisor reported max_int? or went back and forth reporting more and then less again?

Reply:
Currently since we rely on cAdvisor's cache its updated in 5 min interval by default and can be configured via cAdvisor flag `update_machine_info_interval`.
So any fluctuation within this interval does affect kubelet and the final value will be picked by kubelet.

In case of up-size and downsize in two consecutive intervals the Node object will be updated with the latest/actual machine info.

Reply:
ok, but we kinda expect only upsizing events, this is implicit. So should we ever get downsizing events, how do we treat them?

Reply:
Currently we expect only up-size events.

But in-case of a downsize event also, The kubelet will report the updated/actual machine capacity and
In that case , any running pods consuming more than available resources will hit OOMKilled state.

Reply:
I don't understand this statement. So the side effect of this KEP will be that we can start OOM killing pods in case when cadvisor reported lower values? I thought this KEP is only about increasing sizes.

Reply:
Apologies, Let me try to bring more clarity on this regard. 

Yes, Though this KEP is about increasing node capacities, but I wanted to highlight known limitation that exist even in the current implementation incase of downsize of resource.

For example We have a node with 8G memory and pod running on the node consuming 6G, In case of downsize of memory of Node to 4G from the system side of things it still possible that the Pod will hit the OOM killed state.

Reply:
Arguably processing downsizing events IS a form of handling hotunplug resources, so I don't think this is a good solution for this work. I think we need more investigation in this area and perhaps revisit how we fetch the data (cadvisor)

Reply:
IIUC it would be beneficial to compare the new behavior with the current one.

> what about cases when cadvisor reported max_int? or went back and forth reporting more and then less again?

Can this happen today? Don't we just count on cadvisor to report initial valid data, then we simply ignore any changes to its cache? Can't we just ignore downsize events like we already do now?

Sorry if this doesn't make sense.

Reply:
I concur that the case of downsizing needs to be more fully explored/explained. Because that can (and most probably will) happen in practice.

E.g. what happens for guaranteed pods if the cpus (or part of the cpus) they are pinned to go away. Are the managers handling this scenario correctly. What if the `--reserved-cpus` is not part of the new set of cpus.

Reply:
Added a section about Handling HotUnplug Events. As suggested we are planning to have a error mode to report the unplug events.